Implement "Highlight Character, String or Regex" dialog

tangledhelix opened the pull request:
Implements the "Highlight Character, String or Regex" dialog.

Fixes #41 

## Testing notes

Generally this should behave the same as in GG1, with a couple of exceptions. Notably, it co-exists with "highlight alignment column" and "highlight surrounding brackets/quotes", which in GG1 was not the case. We discussed that in some other ticket and decided we would make that change.

Beyond that it should behave like GG1:
- Only one highlight dialog should exist, no duplicate dialogs
- Only highlights within a selection
- Searches (both types) can be case-sensitive or not, controlled with "Match case"
- Toggle Regex mode with checkbox
- Regex should display as red if it won't compile
- Mutually exclusive with "single quote" & "double quote" highlighters; only one can be active at a time
- Clear Highlights button is equivalent to the Remove Highlights menu item
- Restore Selection button should re-select the prior selection
- Regex & Match case checkboxes should persist, including across restarts
- Search history is for this dialog only; not shared with the S/R dialog
- Use same highlight color as "single quote" & "double quote" highlight.
    - Renamed that highlight symbol to MANUAL to group all manually-triggered highlights into one group
- Pressing Return is the same as clicking the Highlight button

## Other notes

I refactored a bit of other stuff for this. For instance the quotes highlighter naively used +1c for highlights, and had to be fixed to work with matches of lengths other than 1. I also renamed the QUOTEMARK highlight to MANUAL.

I added a checkbox "Match Case" that allows the user to control case-sensitivity. This is a divergence from GG1. The default is to match case, which matches the GG1 behavior.

I added the keypress handler for Return, which was not part of the old tool.

I renamed a couple of the buttons from GG1 to be more consistent with menu items (or at least they are consistent in macOS): Select All, Restore Selection

This is a draft for now because it's based on the branch from #556 - but it's ready to test now.

tangledhelix commented on src/guiguts/search.py:
```diff
@@ -623,6 +628,133 @@ def display_message(self, message: str = "") -> None:
+            self.top_frame,
+            text="Previous Selection",
+            takefocus=False,
+            width=14,
```
In general I do agree. If I remember right, the reason these buttons had widths applied is they were not laying out well due to the differing text widths, like this:

<img width="680" alt="Screenshot 2024-11-30 at 10 50 10 AM" src="https://github.com/user-attachments/assets/122f6ad7-9db9-4d62-aa53-018c74a74987">

I'll look for examples in the other dialogs to improve this. For example I notice the buttons in S/R seem to lay out evenly even though no width is defined.

<img width="800" alt="Screenshot 2024-11-30 at 10 52 15 AM" src="https://github.com/user-attachments/assets/44c53dd9-4c9f-4bd3-b93d-f4ef34d509cb">

tangledhelix replied:
I'm guessing `sticky="NSEW"` and the grid definition are the key here; I'll poke at it some more.

windymilla replied:
Yes - if two buttons are in the same column, and both have at least "EW" stickiness, then they should be the same width as one another. I don't think there's a way to force the buttons in the left hand column to be the same width as those in the right, but that's possibly less serious. You can also affect how columns (& hence the buttons if they are "EW" sticky) resize as the dialog is resized by using the `weight` parameter (which you will see used in several places) but I doubt you would want/need to use it here. Columns with the same weight should grow/shrink in a similar manner.

tangledhelix replied:
I think as I submitted this, that's true, although I decided to add a "Match Case" checkbox and now I have columnspan... so it's needed for the columns to balance. Dropping weight from the row definitions seems to have no effect, though, so I'll omit it there.

tangledhelix replied:
The layout should be a lot more sensible now, and scales dynamically (horizontal/x-axis resize only), and the geometry should persist across sessions.

tangledhelix commented on src/guiguts/search.py:
```diff
@@ -623,6 +628,133 @@ def display_message(self, message: str = "") -> None:
+        style.configure("BadRegex.TCombobox", foreground=new_col)
+
+        def is_valid_regex(new_value: str) -> bool:
+            """Validation routine for Search Combobox - check value is a valid regex.
```
This should work properly now. I also changed from radio buttons to a Regex checkbox, which is more consistent with elsewhere in GG, like the S/R dialog.

tangledhelix commented on src/guiguts/search.py:
```diff
@@ -623,6 +628,133 @@ def display_message(self, message: str = "") -> None:
+        kwargs["resize_y"] = False
+        kwargs["disable_geometry_save"] = True
+        super().__init__("Highlight Character(s) or Regex", *args, **kwargs)
+        self.minsize(350, 115)
```
Everything should be dynamically sized and scale sanely now.

tangledhelix commented:
I pushed a new version, addressing the previous remarks. I also did a few new things.

- Pressing enter/return is the same as clicking the Highlight button, which is now styled as "active"
- Renamed the buttons to be more consistent within the app, or just more clear
- Added "Match case" to control case-sensitivity. Defaults to on, which is the GG1 behavior.
- Changed Exact/Regex radio button pair to a single checkbox for Regex.

Ready for re-test / review.

windymilla reviewed (approved):
Layout much nicer, thanks. 
Regex & Match Case checkboxes working correctly.

srjfoo commented:
It looks like (at least for me) `\b` doesn't work in regex mode for word boundaries.

windymilla commented:
> It looks like (at least for me) `\b` doesn't work in regex mode for word boundaries.

Suggestion made via Slack

tangledhelix commented:
Per discussion in Slack, will not copy this from GG1 to GG2 for now. There is a lot of overlap with the existing Search dialog and Find All, etc. We can revisit at a later date if needed.
